nipype/pipeline/plugins/multiproc.py

```diff
@@ -134,207 +114,190 @@ def __init__(self, plugin_args=None):
 
-        # Check to see if a job is available
-        currently_running_jobids = np.flatnonzero((self.proc_pending == True) & \
-                                (self.depidx.sum(axis=0) == 0).__array__())
+        # Check all jobs without dependency not run
+        jobids = np.flatnonzero(
+            ~self.proc_done & (self.depidx.sum(axis=0) == 0).__array__())
 
         # Check available system resources by summing all threads and memory used
-        busy_memory_gb = 0
-        busy_processors = 0
-        for jobid in currently_running_jobids:
-            if self.procs[jobid]._interface.estimated_memory_gb <= self.memory_gb and \
-                            self.procs[jobid]._interface.num_threads <= self.processors:
-
-                busy_memory_gb += self.procs[jobid]._interface.estimated_memory_gb
-                busy_processors += self.procs[jobid]._interface.num_threads
+        free_memory_gb, free_processors = self._check_resources(self.pending_tasks)
 
-            else:
-                raise ValueError(
-                    "Resources required by jobid {0} ({3}GB, {4} threads) exceed what is "
-                    "available on the system ({1}GB, {2} threads)".format(
-                        jobid, self.memory_gb, self.processors,
-                        self.procs[jobid]._interface.estimated_memory_gb,
-                        self.procs[jobid]._interface.num_threads))
+        logger.info('Currently running %d tasks, and %d jobs ready. '
+                     'Free memory (GB): %0.2f/%0.2f, Free processors: %d/%d',
+                     len(self.pending_tasks), len(jobids),
+                     free_memory_gb, self.memory_gb, free_processors, self.processors)
 
-        free_memory_gb = self.memory_gb - busy_memory_gb
-        free_processors = self.processors - busy_processors
 
-        # Check all jobs without dependency not run
-        jobids = np.flatnonzero((self.proc_done == False) & \
-                                (self.depidx.sum(axis=0) == 0).__array__())
+        if (len(jobids) + len(self.pending_tasks)) == 0:
+            logger.debug('No tasks are being run, and no jobs can '
+                         'be submitted to the queue. Potential deadlock')
+            return
 
         # Sort jobs ready to run first by memory and then by number of threads
         # The most resource consuming jobs run first
-        jobids = sorted(jobids,
-                        key=lambda item: (self.procs[item]._interface.estimated_memory_gb,
-                                          self.procs[item]._interface.num_threads))
-
-        if str2bool(config.get('execution', 'profile_runtime')):
-            logger.debug('Free memory (GB): %d, Free processors: %d',
-                         free_memory_gb, free_processors)
+        # jobids = sorted(jobids,
+        #                 key=lambda item: (self.procs[item]._get_mem_gb(),
+        #                                   self.procs[item]._get_n_procs()))
```

Comment:
Should this be uncommented or removed?

Reply:
Thanks for pointing out. This was actually something I want to discuss with you, @satra and it probably may attract @chrisfilo's interest.

Do we really want to sort nodes by size? I'd say we don't (and I'd remove this commented code)

Reply:
I think the idea is to fill as many slots as possible, and let the small nodes fill in the holes. Which makes sense; if I have something that's going to use half the cores and take a while, and a bunch of small jobs that can run more-or-less serially in the remaining cores, we'll use the processor pretty efficiently. (And actually would mean we should be sorting in the opposite order; threads then memory.)

One metric I've thought about is the number of nodes that will be able to be queued when a given node is completed. The priority then would be to remove bottlenecks first, and fill in around them.

Scheduling priority metric might be a sensible plugin argument. And the default could be `tsort` (graph as it comes in) or `mem_thread` (current strategy), and we could easily build alternatives like un-bottleneck...

Reply:
@oesteban, @effigies, and @ccraddock - my first thought is that we shouldn't get into the scheduling business - as that is a hairy one, even on a multi-core machine. that being said i'm ok with some basic scheduling strategies. since we don't know how long something is going to run (at least yet), the best we can do is prioritize cores, so favor small jobs over long jobs. however, since we have a graph, the other piece would be how many downstream jobs are dependent on the current node. as i write this, i realize why my first thought was first.

Reply:
I mean, we're already in the scheduling business, at least insofar as there's topological sorting for all other plugins, and memory-then-threads for `MultiProc`.

The interface could be dead simple:

```Python
def mem_thread_scheduler(jobids, plugin):
    return sorted(jobids, key=lambda item: (plugin.procs[item]._get_mem_gb(),
                                            plugin.procs[item]._get_n_procs()))

def tsort(jobids, plugin):
    return jobids
```

Reply:
@effigies - we are only partially in the scheduling business :) we hand of things from the graph. while the simple interface would work, it's a starting point, as this doesn't take the weight of dependencies or their run profiles into account.

this is almost like playing chess :) peeking a few more levels may give us an A* algorithm.

but i would be ok with this as a start. however, more generally this could be applied to any of the scheduler plugins (except the graph ones).

Reply:
Since it was already in the codebase, I thought that we could allow users to try this "mem_thread" scheduler. To do so, `MultiProc` allows now a `plugin_args['scheduler'] = 'mem_threads'` argument. However, I do agree with @satra that scheduling is a pretty hairy problem (and worse when you don't have a wallclock time).